src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MeanQualityByCycleSpark.java

```diff
@@ -0,0 +1,273 @@
+            return this;
+        }
+
+        Histogram<Integer> getMeanQualityHistogram() {
+            final String label = useOriginalQualities ? "MEAN_ORIGINAL_QUALITY" : "MEAN_QUALITY";
+            final Histogram<Integer> meanQualities = new Histogram<>("CYCLE", label);
+
+            int firstReadLength = 0;
+
+            for (int cycle=0; cycle < firstReadTotalsByCycle.length; ++cycle) {
+                if (firstReadTotalsByCycle[cycle] > 0) {
+                    meanQualities.increment(cycle, firstReadTotalsByCycle[cycle] / firstReadCountsByCycle[cycle]);
+                    firstReadLength = cycle;
+                }
+            }
+
+            for (int i=0; i< secondReadTotalsByCycle.length; ++i) {
+                if (secondReadCountsByCycle[i] > 0) {
+                    final int cycle = firstReadLength + i;
+                    meanQualities.increment(cycle, secondReadTotalsByCycle[i] / secondReadCountsByCycle[i]);
+                }
+            }
+
+            return meanQualities;
+        }
+
+        boolean isEmpty() {
+            return maxLengthSoFar == 0;
+        }
+
+    }
+
+    /**
+     * This pair makes it easy to process the set of reads once and maintain both histograms.
+     */
+    static final class HistogramGeneratorPair implements Serializable{
+
+        private static final long serialVersionUID = 1;
+        HistogramGenerator q = new HistogramGenerator(false);
+        HistogramGenerator oq = new HistogramGenerator(true);
+        HistogramGeneratorPair addRead(GATKRead read){
+            q.addRecord(read);
+            oq.addRecord(read);
+            return this;
+        }
+        HistogramGeneratorPair merge(HistogramGeneratorPair other){
+            q.merge(other.q);
+            oq.merge(other.oq);
+            return this;
+        }
+
+    }
+
+    @Override
+    protected void runPipeline(final JavaSparkContext ctx) {
+        if ( readArguments.getReadFilesNames().size() != 1 ) {
+            throw new UserException("This tool only accepts a single bam/sam/cram as input");
+        }
+        final String bam = readArguments.getReadFilesNames().get(0);
+        final ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        final SAMFileHeader readsHeader = ReadsSparkSource.getHeader(ctx, bam);
+        setup(readsHeader);
+
+        /**
+         * If no intervals specified - use all reads, mapped and unmapped.
+         */
+        final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(readsHeader.getSequenceDictionary())
+                : null;
+
+        final JavaRDD<GATKRead> reads = readSource.getParallelReads(bam, intervals);
+        final JavaRDD<GATKRead> filteredReads = reads.filter(makeReadFilter(PF_READS_ONLY, ALIGNED_READS_ONLY));
+        final HistogramGeneratorPair aggregate = filteredReads.aggregate(new HistogramGeneratorPair(),
+                (hgp, read) -> hgp.addRead(read),
```

Comment:
i'm fine either way. i think explicit variables like this is a bit more readable